scripts/2.TCGA-MLexample-covariates.py

```diff
@@ -80,108 +79,112 @@
+        'features__expressions__pca__n_components' : n_components_list,
+        'classify__loss': ['log'],
+        'classify__penalty': ['elasticnet'],
+        'classify__alpha': [10 ** x for x in range(-3, 1)],
+        'classify__l1_ratio': [0, 0.2, 0.8, 1],
+    },
+    'covariates': {
+        'classify__loss': ['log'],
+        'classify__penalty': ['elasticnet'],
+        'classify__alpha': [10 ** x for x in range(-3, 1)],
+        'classify__l1_ratio': [0, 0.2, 0.8, 1],
+    }
 }
 
-pipeline = Pipeline(steps=[
+# Feature selection
+expression_features = Pipeline([
+    ('select_features', FunctionTransformer(select_feature_set_columns,
+        kw_args={'feature_set': 'expressions'})),
     ('standardize', StandardScaler()),
-    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+    ('pca', PCA())
+])
+covariate_features = Pipeline([
+    ('select_features', FunctionTransformer(select_feature_set_columns,
+        kw_args={'feature_set': 'covariates'})),
+    ('standardize', StandardScaler())
 ])
 
+# Full model pipelines
+pipeline_definitions = {
+    'full': Pipeline([
+        ('features', FeatureUnion([
+            ('expressions', expression_features),
+            ('covariates', covariate_features)
+        ])),
+        ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+    ]),
+    'expressions': Pipeline([
+        ('features', FeatureUnion([('expressions', expression_features)])),
+        ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+    ]),
+    'covariates': Pipeline([
+        ('features', FeatureUnion([('covariates', covariate_features)])),
+        ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+    ])
+}
+
 models = ['full', 'expressions', 'covariates']
 
 cv_pipelines = {mod: GridSearchCV(estimator=pipeline, 
-                             param_grid=param_grid, 
+                             param_grid=param_grids[mod], 
                              n_jobs=1, 
-                             scoring='roc_auc') for mod in models}
+                             scoring='roc_auc') 
+                for mod, pipeline in pipeline_definitions.items()}
 
 
-# In[15]:
+# In[12]:
 
-get_ipython().run_cell_magic('time', '', "for model, pipeline in cv_pipelines.items():\n    print('Fitting CV for model: {0}'.format(model))\n    pipeline.fit(X=X_partitions.get(model).get('train'), y=y_train)\n# cv_pipeline_full.fit(X=X_train_full, y=y_train)")
+get_ipython().run_cell_magic('time', '', "for model, pipeline in cv_pipelines.items():\n    print('Fitting CV for model: {0}'.format(model))\n    pipeline.fit(X=X_train, y=y_train)")
 
 
-# In[16]:
+# In[13]:
 
-# Best Params
+# Best Parameters
 for model, pipeline in cv_pipelines.items():
     print('{0}: {1:.3%}'.format(model, pipeline.best_score_))
 
-    # Best Params
     print(pipeline.best_params_)
 
 
 # ## Visualize hyperparameters performance
 
-# In[17]:
+# In[14]:
 
 cv_results_df_dict = {model: 
```

Comment:
I think this should be a single dataframe with a `model` column. Then you can do 
```
cv_results_df.query("@model == 'full'")
```

below (if we keep the grid viz cell below.)